File: commands/echo-json.rb
#!/usr/bin/env ruby

require 'json'
require_relative '../sources/rb.rb'

class HelpError < AbstractHelpError
  def initialize(message = "")
    super(message)
    self.help = [
      'USAGE:',
      'echo-json.rb <make> [--] ...[<key> <value>]',
      'echo-json.rb <stringify|encode|decode|json|pretty> [--] ...<input>',
    ].join("\n")
  end
end

def real(input)
  begin
    JSON.parse(input)
  rescue
    input
  end
end

def parse_json(input)
  begin
    { parseError: nil, value: JSON.parse(input) }
  rescue => parseError
    { parseError: parseError, value: input }
  end
end

def main(*args)
  # parse <action>
  if args.length == 0
    raise HelpError.new("No <action> was provided.")
  end
  actions = ['make', 'stringify', 'encode', 'decode', 'json', 'pretty']
  unless actions.include?(args[0])
    raise HelpError.new("An unrecognised <action> was provided: #{args[0]}")
  end
  action = args.shift
  # parse [...options] ...<input>
  properties = []
  inputs = []
  while args.length > 0
    arg = args.shift
    if arg == '--'
      inputs.concat(args)
      break
    elsif arg.start_with?('--property=')
      properties << arg.sub(/^--property=/, '')
    else
      inputs << arg
    end
  end
  # <make> ...[<key> <value>]
  if action == 'make'
    # validate we have a <value> for every <key>
    if inputs.length % 2 != 0
      raise '<make> requires an even number of <key> <value> pairs.'
    end
    # build the object
    output = {}
    while inputs.length > 0
      key = inputs.shift
      value = inputs.shift
      output[key] = real(value)
    end
    return write_stdout_stringify(output)
  end
  # <action> ...<input>
  while inputs.length > 0
    input = inputs.shift
    case action
    when 'stringify'
      write_stdout_stringify(input)
    when 'json', 'pretty'
      parsed = parse_json(input)
      if parsed[:parseError] != nil
        raise "Failed to parse what should be JSON-encoded <input> = #{parsed[:parseError].message}"
      end
      if action == 'json'
        write_stdout_stringify(parsed[:value])
      else
        write_stdout_pretty(parsed[:value])
      end
    when 'encode'
      parsed = parse_json(input)
      write_stdout_stringify(parsed[:value])
    when 'decode'
      parsed = parse_json(input)
      if properties.length == 0
        output = if parsed[:value].is_a?(Hash) || parsed[:value].is_a?(Array) || parsed[:value].nil?
                   JSON.generate(parsed[:value])
                 else
                   parsed[:value].to_s
                 end
        write_stdout_plain(output)
      else
        outputs = []
        properties.each do |property|
          keys = property.split('.')
          diver = parsed[:value]
          keys.each do |key|
            if diver.is_a?(Hash) && diver.key?(key)
              diver = diver[key]
            else
              raise "Property \"#{property}\" does not exist in the provided input."
            end
          end
          if diver.is_a?(Hash) || diver.is_a?(Array) || diver.nil?
            outputs << JSON.generate(diver)
          else
            outputs << diver.to_s
          end
        end
        write_stdout_plain(outputs.join("\n"))
      end
    when 'parse'
      parsed = parse_json(input)
      if parsed[:parseError] != nil
        raise "Failed to parse what should be JSON-encoded <input> = #{parsed[:parseError].message}"
      end
      write_stdout_pretty(parsed[:value])
    else
      raise "Invalid action: #{action}"
    end
  end
end

begin
  main(*ARGV)
rescue => error
  exit_with_error(error)
end

File: commands/echo-json.ts
#!/usr/bin/env -S eval-wsl deno run --quiet --no-config --no-lock --no-npm --no-remote --cached-only
// deno-lint-ignore-file no-unreachable

import {
	HelpError,
	exitWithError,
	writeStdoutStringify,
	writeStdoutPlain,
	writeStdoutPretty,
	heredoc,
	wantsHelp,
	asString,
	asError,
	assertFactory,
} from '../sources/ts.ts'

const { values: actionValues, as: asAction } = assertFactory(
	['make', 'stringify', 'encode', 'decode', 'json', 'pretty'] as const,
	'action',
)

class UsageError extends HelpError {
	override help = heredoc`
		USAGE:
		\`echo-json.ts <make> [--] ...[<key> <value>]\`
		\`echo-json.ts <${actionValues.join('|')}> [--] ...<input>\`
	`
}

// Parsing
function real(input: string) {
	try {
		return JSON.parse(input)
	} catch {
		return input
	}
}
function parse(input: string): {
	parseError: Error | null
	value: unknown
} {
	try {
		return {
			parseError: null,
			value: JSON.parse(input),
		}
	} catch (parseError) {
		return {
			parseError: asError(parseError),
			value: input,
		}
	}
}

// Execute
// <action> [...options] ...<input>
function main(...args: string[]) {
	// parse <action>
	if (args.length === 0) {
		throw new UsageError(`--help=No <action> was provided.`)
	}
	const action = asAction(args.shift())
	// parse [...options] ...<input>
	const properties = [],
		inputs = []
	while (args.length !== 0) {
		const input = asString(args.shift())
		if (input == '--') {
			inputs.push(...args)
			break
		} else if (input.startsWith('--property=')) {
			const property = input.replace(/^--property=/, '')
			properties.push(property)
		} else {
			inputs.push(input)
		}
	}
	// <make> ...[<key> <value>]
	if (action === 'make') {
		// validate we have a <value> for every <key>
		if (inputs.length % 2 !== 0) {
			throw new UsageError(
				`--help=<make> requires an even number of <key> <value> pairs.`,
			)
		}
		// build the object
		const output: Record<string, unknown> = {}
		while (inputs.length) {
			const key = asString(inputs.shift())
			const value = asString(inputs.shift())
			output[key] = real(value)
		}
		return writeStdoutStringify(output)
	}
	// ...<input>
	while (inputs.length) {
		const input = asString(inputs.shift())
		switch (action) {
			case 'stringify': {
				return writeStdoutStringify(input)
				break
			}
			case 'json':
			case 'pretty': {
				const { parseError, value } = parse(input)
				if (parseError != null) {
					throw new UsageError(
						`--help=Failed to parse what should be JSON-encoded <input> = `,
						`--value=${parseError.message}`,
					)
				}
				if (action == 'json') {
					return writeStdoutStringify(value)
				} else {
					return writeStdoutPretty(value)
				}
				break
			}
			case 'encode': {
				const { value } = parse(input)
				return writeStdoutStringify(value)
				break
			}
			case 'decode': {
				const { value } = parse(input)
				let output = ''
				if (properties.length === 0) {
					output =
						typeof value === 'object' ? JSON.stringify(value) : String(value)
				} else {
					const outputs = []
					for (const property of properties) {
						const keys = property.split('.')
						let diver: any = value
						for (const key of keys) {
							if (diver && typeof diver === 'object' && key in diver) {
								diver = diver[key]
							} else {
								throw new UsageError(
									`Property `,
									`--value=${property}`,
									` does not exist in the provided input.`,
								)
							}
						}
						outputs.push(
							typeof diver === 'object' ? JSON.stringify(diver) : String(diver),
						)
					}
					output = outputs.join('\n')
				}
				return writeStdoutPlain(output)
				break
			}
			default:
				throw new UsageError(`--help=Invalid <action>: `, `--value=${action}`)
				break
		}
	}
}

try {
	if (wantsHelp(Deno.args)) {
		throw new UsageError()
	}
	await main(...Deno.args)
} catch (error) {
	await exitWithError(error)
}
